tasks/image/build.go

```diff
@@ -75,7 +78,65 @@ func buildIsStale(ctx *context.ExecuteContext, t *Task) (bool, error) {
 }
 
 func buildImage(ctx *context.ExecuteContext, t *Task) error {
-	if err := Stream(os.Stdout, func(out io.Writer) error {
+	var err error
+	switch t.hasContent() {
+	case true:
+		err = t.runContainerFromTarBall(ctx)
+	default:
+		err = t.runContainerFromDockerfile(ctx)
+	}
+	if err != nil {
+		return err
+	}
+	image, err := GetImage(ctx, t.config)
+	if err != nil {
+		return err
+	}
+	record := imageModifiedRecord{ImageID: image.ID}
+	return updateImageRecord(recordPath(ctx, t.config), record)
+}
+
+func buildArgs(args map[string]string) []docker.BuildArg {
+	out := []docker.BuildArg{}
+	for key, value := range args {
+		out = append(out, docker.BuildArg{Name: key, Value: value})
+	}
+	return out
+}
+
+func (t *Task) hasContent() bool {
+	if len(t.config.Steps) != 0 {
+		return true
+	}
+	return false
+}
+
+func (t *Task) writeTarball() (*bytes.Buffer, error) {
+	inputbuf := bytes.NewBuffer(nil)
+	rightNow := time.Now()
+	tr := tar.NewWriter(inputbuf)
+	header := &tar.Header{Name: "Dockerfile",
+		Size:       t.getContentSize(),
+		ModTime:    rightNow,
+		AccessTime: rightNow,
+		ChangeTime: rightNow,
+	}
```

Comment:
I think this is the problem. The tarball needs to contain all the files in the Context directory, not just the Dockerfile.

We should try to re-use some code from `docker/docker` for this, so that it honors the `.dockerignore` file correctly.

The content size should come from the tarball size, not the size of the steps.